sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,236 @@
+                {
+                    throw OperationFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(Value));
+                }
+
+                _value = value;
+                HasValue = true;
+            }
+        }
+
+        public Response RawResponse { get; set; }
+
+        public string OperationTypeName { get; set; }
+
+        public TimeSpan DefaultPollingInterval { get; set; }
+
+        protected ClientDiagnostics ClientDiagnostics { get; }
+
+        protected RequestFailedException OperationFailedException { get; set; }
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                var pollingInterval = response.Headers.TryGetValue(RetryAfterHeaderName, out string retryAfterValue)
+                    && int.TryParse(retryAfterValue, out int delayInSeconds)
+                    ? TimeSpan.FromSeconds(delayInSeconds) : DefaultPollingInterval;
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        {
+            Response response = async
+                ? await _operation.GetResponseAsync(cancellationToken).ConfigureAwait(false)
+                : _operation.GetResponse(cancellationToken);
+
+            var state = _operation.UpdateState(response);
```

Comment:
We call GetResponse and then immediately the UpdateState. Is it worth merging into one method?

Reply:
I have update only Form Recognizer and Text Analytics with the suggested approach for comparison. It doesn't change that much from the concrete Operation side, but the OperationInternal implementation gets incredibly simpler.

Since we don't need the `TResponseType` generic type anymore, we can completely eliminate the extra interface `IOperationInternal<TResult, TResponseType>` and its concrete implementation.

For this reason, I'm totally in favor of merging the methods into a single one.